base/src/verifiable.rs

```diff
@@ -0,0 +1,16 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+//! Traits and helpers for data structures that require validation
+
+/// Types that implement Verifiable are able to self-check using intrinsic data (e.g. signatures)
+/// and optionally some context that is required for peforming the verification
+/// e.g. a singleton signature verifier or param required for snark verification
+pub trait Verifiable<CTX> {
+    /// Accepts self and a context and
+    /// returns if the implementor is valid
+    ///
+    /// The context must be a mutable reference to satisfy the requirements of using a
+    /// shared signature checking context which performs self mutation as an optimization
+    fn verify(&self, ctx: &mut CTX) -> bool;
```

Comment:
Can the verify operation fail. If yes, probably good to return a `Result` with error context.

Reply:
I considered this but I think that a failure is just something not being valid. Also returning  a bool keeps it consistent with the proof-systems `verify` function https://github.com/o1-labs/proof-systems/blob/3961f5cfce1f8e2f014ff8afeb50241eb6a161a4/signer/src/schnorr.rs#L71

Reply:
sounds good